clean conversion, one commit to rxjava2

vladp opened the pull request:
Hello,
I had performed an update to convert everything to rxjava2.
I had to do it for my own needs, and wanted to share.

I made sure the changes are clean and easy to review in a pull request.
I tested with API 22.

Regards,
Vladislav

Commit:
Merge pull request #4 from athkalia/task/test

test

Commit:
Update README.md

Commit:
Merge pull request #5 from athkalia/athkalia-patch-1

Update README.md

Commit:
Update README.md

Commit:
Merge pull request #6 from athkalia/athkalia-patch-1

Update README.md

Commit:
clean conversion, one commit to rxjava2

athkalia commented:
Hey there!
thanks for the PR. I'll review over the weekend and get back to you : >

miguelslemos reviewed (commented):
looks good

vladp commented:
Hi, it says above 'TeamCity build failed'. but when I click on details link, I get
http://localhost:9091/viewLog.html?buildId=456&buildTypeId=JustAnotherAndroidApp_PullRequestJob
(which I cannot access).

how can I see what the problem is?

I remember that when converting, rxjava2 there is no such thing as 'immediate' for Schedulers (it apparently did not work right in rxjava1 either)
So I just fed into the test case same schedulers as the asserts where expecting .
May be that's a problem?

thx

athkalia commented:
Hey pal. I am running a locally hosted teamcity set up that is unavailable at the moment unfortunately as I don't have a static ip.  The problem is with the PMD static analysis tool. If you run the command "gradlew pmd" a report will be generated in the build folder that contains what needs fixing. I had a look, and it's an unused import in the MainPresenterTest class.

The thing is that before upgrading the project to rxJava 2 I wanted to have a quick look on rxJava 2 myself but unfortunately I didn't find the time over the weekend. As soon as I do that, I will happily accept the PR.
One more thing, if possible, can you sort out the merge conflicts, as I did a push earlier today that seems to have broken something in your PR.

Commit:
removed unused import

vladp commented:
Hello.  I removed the unused import and ran "gradlew pmd" to make sure it completed successfully.

I synched up my fork with your upstream, removed the unused import, and push back into my develop branch.
Also saw the merge conflict, updated that as well.

Commit:
Merge remote-tracking branch 'upstream/master' into develop

Commit:
merged develop branch for original.

athkalia commented on app/build.gradle:
```diff
@@ -234,11 +234,15 @@ dependencies {
+
     // Rx
-    compile "io.reactivex:rxandroid:$versions.rxAndroidVersion"
+    compile "io.reactivex.rxjava2:rxandroid:$versions.rxAndroidVersion" //rxjava2
```
can you please remove trailing comment "//rxjava2"

vladp replied:
done

athkalia commented on app/build.gradle:
```diff
@@ -234,11 +234,15 @@ dependencies {
     // Already has RxJava, but still better to explicitly get latest version.
-    compile "io.reactivex:rxjava:$versions.rxJavaVersion"
-    compile "com.squareup.retrofit2:adapter-rxjava:$versions.retrofitVersion"
+    compile "io.reactivex.rxjava2:rxjava:$versions.rxJavaVersion"  //rxjava2
```
same here

vladp replied:
done

athkalia commented on app/build.gradle:
```diff
@@ -234,11 +234,15 @@ dependencies {
-    compile "io.reactivex:rxjava:$versions.rxJavaVersion"
-    compile "com.squareup.retrofit2:adapter-rxjava:$versions.retrofitVersion"
+    compile "io.reactivex.rxjava2:rxjava:$versions.rxJavaVersion"  //rxjava2
+    compile "com.squareup.retrofit2:adapter-rxjava2:$versions.retrofitVersion" //rxjava
```
same here

vladp replied:
done

athkalia commented on app/build.gradle:
```diff
@@ -234,11 +234,15 @@ dependencies {
-    compile "com.squareup.retrofit2:adapter-rxjava:$versions.retrofitVersion"
+    compile "io.reactivex.rxjava2:rxjava:$versions.rxJavaVersion"  //rxjava2
+    compile "com.squareup.retrofit2:adapter-rxjava2:$versions.retrofitVersion" //rxjava
+    //0.16.0" //for assemblyTracking in debug
```
can you please remove the '' //0.16.0"  part of the comment.
Also, do you happen to have any links on whether they are planning to add this feature back into the official library or why they removed it in the first place?

vladp replied:
done

athkalia commented on app/config/gradle/versioning.gradle:
```diff
@@ -26,17 +26,18 @@ ext.versions = [
+        retrofitVersion           : '2.2.0',
+        rxJavaVersion             : '2.0.7',
+        rxAndroidVersion          : '2.0.1',
+        rxjava2ExtenionsVersion   : '0.16.0',
```
you have a typo, should be rxJava2ExtensionsVersion

athkalia replied:
Also, latest version is '0.16.2'

athkalia commented on app/src/integrationTests/java/com/example/networking/RestServiceIntegrationTest.java:
```diff
@@ -38,7 +38,7 @@ public void afterEveryTest() throws IOException {
         mockWebServerHelper.enqueueJsonResponseFromFileForMockWebServer(mockWebServer, "shots_response.json");
 
-        List<ShotResponse> shotResponseList = restService.getShots().toBlocking().value();
+        List<ShotResponse> shotResponseList = restService.getShots().toObservable().blockingSingle();
```
can't we use the 'blockingGet()' operator?

vladp replied:
blockingGet does not work on continuous stream. I was not sure, tbh, if the testing model involves continuous stream or just a small block with a clear terminate.

athkalia replied:
I don't quite understand what you are talking about. What do you mean by continuous stream? This is just a Single<List<>>

athkalia commented on app/src/main/java/com/example/features/dashboard/presenter/MainPresenter.java:
```diff
@@ -48,7 +49,7 @@ public void fetchShots() {
                 .doAfterTerminate(countingIdlingResource::decrement)
                 .subscribeOn(rxSchedulers.getIoScheduler())
                 .toObservable()
```
we can combine .toObservable() and the next line and replace them with
```
                .flatMapObservable(Observable::fromIterable)
```

vladp replied:
done

athkalia commented on app/src/unitTests/java/com/example/features/dashboard/presenter/MainPresenterTest.java:
```diff
@@ -49,7 +53,7 @@
     public void setUp() throws Exception {
         MockitoAnnotations.initMocks(this);
-        RxSchedulers rxSchedulers = new RxSchedulers(Schedulers.immediate(), Schedulers.immediate(), Schedulers.immediate(), Schedulers.immediate());
+        RxSchedulers rxSchedulers = new RxSchedulers(Schedulers.io(), Schedulers.computation(), Schedulers.trampoline(), AndroidSchedulers.mainThread());
```
the RxSchedulers class acts as a wrapper that I can inject with the immediate() schedulers for tests. 
In rxjava 1 there's another way to handle this that looks like this:   
```
RxJavaHooks.setOnIOScheduler(scheduler -> Schedulers.immediate());
```
to change a specific Scheduler.

In rxjava 2 it's moved to the RxJavaPlugins class:

```
RxJavaPlugins.setIoSchedulerHandler(scheduler -> Schedulers.trampoline());
```

(note that immediate() scheduler no longer exists.)

Just pointing this out, you don't need to make this change, I'll do it in the very near future.

The only thing you need to do to make the tests pass it to change the line 

```
  RxSchedulers rxSchedulers = new RxSchedulers(Schedulers.immediate(), Schedulers.immediate(), Schedulers.immediate(), Schedulers.immediate());
```

to 

```
        RxSchedulers rxSchedulers = new RxSchedulers(Schedulers.trampoline(), Schedulers.trampoline(), Schedulers.trampoline(), Schedulers.trampoline());
```
and I'll sort out the rest in another PR

vladp replied:
done

athkalia commented on app/src/unitTests/java/com/example/features/dashboard/presenter/MainPresenterTest.java:
```diff
@@ -49,7 +53,7 @@
+        RxSchedulers rxSchedulers = new RxSchedulers(Schedulers.io(), Schedulers.computation(), Schedulers.trampoline(), AndroidSchedulers.mainThread());
         mainPresenter = new MainPresenter(mockRestService, rxSchedulers, mockAnalyticsHelper, new ShotMapper(), new CountingIdlingResource("test"));
         mainPresenter.attachView(mockMainView);
     }
```
there's another issue with this class: One of the tests is failing. Test with name:
"null_shots_are_filtered_before_returned"

This is because in the second line of the test there is a 
```
        shotResponseList.add(null);
```

The test seems to be failing because in the MainPresenter 
```
Observable::fromIterable
```
has a null check, i.e. in RxJava 2 null values are no longer allowed (i think)

So to sum up:


a)
in the MainPresenter please change line

```
                .filter(shot -> shot != null && shot.getImagesData() != null && shot.getImagesData().getTeaserImageUrl() != null)

```

to 
```
                .filter(shot -> shot.getImagesData() != null && shot.getImagesData().getTeaserImageUrl() != null)
```

in this test "null_shots_are_filtered_before_returned"

please delete the line 

```
        shotResponseList.add(null);
```

vladp replied:
done

Commit:
updated as per athkalia feedback

athkalia commented:
thanks for the PR pal

vladp commented:
you are welcome. Thank you for providing a good modern app template for the community, and merging my small updates.
